### PR TITLE
Add MQTT dependencies, settings UI, and test coverage (CATROID-1670, CATROID-1672)

### PR DESCRIPTION
**Pull request by Paras-ydv**
## Summary

This PR introduces the initial MQTT foundation for Catroid by combining:

- [CATROID-1670](https://catrobat.atlassian.net/browse/CATROID-1670): Eclipse Paho MQTT dependency and Android configuration
- [CATROID-1672](https://catrobat.atlassian.net/browse/CATROID-1672): MQTT settings UI, persistence layer, validation, and automated tests

#### catroid/build.gradle

- Added `org.eclipse.paho.client.mqttv3:1.2.5`
- Added `org.eclipse.paho.android.service:1.1.1`
- Added `FEATURE_MQTT_ENABLED` build config flag

#### AndroidManifest.xml

- Registered `org.eclipse.paho.android.service.MqttService`
- Verified existing permissions:
  - INTERNET
  - ACCESS_NETWORK_STATE

### Testing

Added dependency sanity tests verifying:

- MqttClient
- MqttConnectOptions
- MqttMessage
- MqttException
- MqttCallback

are available on the test classpath.

### UI Demo

[Watch UI Demo](https://drive.google.com/file/d/1yLadrlKaI-Id3SxY3azNVey9TVCfbOPD/view?usp=sharing)

### Features Implemented

Added MQTT settings screen with:

- MQTT enable toggle
- Broker host
- Broker port
- TLS option
- Username
- Password
- Client ID

### Persistence

Implemented persistent storage using SharedPreferences.

Added accessor APIs:

- isMqttSharedPreferenceEnabled()
- getMqttHost()
- getMqttPort()
- isMqttTlsEnabled()
- getMqttUsername()
- getMqttPassword()
- getMqttClientId()

### Password Storage Refactor (MqttPasswordRepository)

Following a discussion with Harsh and Prof. Wolfgang's review feedback, the MQTT password is no longer stored in plain SharedPreferences.

Instead, a `MqttPasswordRepository` interface was introduced (modelled after `LocalHashVersionRepository`) with a
`DefaultMqttPasswordRepository` implementation backed by `EncryptedSharedPreferences`. It is registered as a Koin singleton and injected wherever the password is read or written.

This keeps `SettingsFragment.getMqttPassword(Context)` as the existing public API — it now delegates to the injected repository internally. Unit tests mock the interface via a Koin test module, so no Android
Keystore is required in the test environment.

### Validation

- Port range: 1–65535
- Invalid values rejected with UI feedback

### Testing

Added automated tests covering:

- Default values
- Persistence
- Validation
- Edge cases
- UI integration

28 unit tests added.

Followed TDD workflow:

- Red
- Green
- Refactor

#### Your Checklist
- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Branch-and-Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

[CATROID-1670]: https://catrobat.atlassian.net/browse/CATROID-1670?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CATROID-1672]: https://catrobat.atlassian.net/browse/CATROID-1672?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    companion object {
+        @JvmField val TAG: String = MqttSettingsFragment::class.java.simpleName
+
+        @VisibleForTesting
+        @JvmStatic
+        fun isValidPort(value: String): Boolean {
+            val port = value.toIntOrNull() ?: return false
+            return port in 1..65535
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt]: @Paras-ydv Please fix all the warnings. Thanks!

[the remark this answers is not in the document]
Comment (Paras-ydv) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt]: Done. Please Check.

File: catroid/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -267,6 +263,7 @@
         <service android:name=".transfers.MediaDownloadService" />
         <service android:name=".utils.notifications.StatusBarNotificationManager$NotificationActionService" />
         <service android:name=".cast.CastService"/>
+        <service android:name="org.eclipse.paho.android.service.MqttService" />
```
Comment (harshsomankar123-tech) [inline catroid/src/main/AndroidManifest.xml]: Do we really need the extra `org.eclipse.paho.android.service` dependency here, or can we keep only the simpler `org.eclipse.paho.client.mqttv3` Java client?

`org.eclipse.paho.android.service` looks outdated/unmaintained since 2017, and Android 12+ restricts starting foreground services from the background. If this wrapper starts a service at the wrong time, Android can throw `ForegroundServiceStartNotAllowedException`.
see :https://developer.android.com/develop/background-work/services/fgs/restrictions-bg-start
WDYT??

File: catroid/build.gradle [hunk 3]
```diff
@@ -423,6 +421,10 @@ dependencies {
     implementation 'com.google.guava:guava:28.2-android'
     implementation 'com.google.code.gson:gson:2.8.7'
 
+    // MQTT
+    implementation 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.5'
+    implementation 'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1'
```
Comment (harshsomankar123-tech) [inline catroid/build.gradle]: We can remove this declaration once we drop the paho.android.service dependency.

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java
Comment (harshsomankar123-tech) [inline catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java:217]: We should add a check for `BuildConfig.FEATURE_MQTT_ENABLED` here and remove the `MQTT_SCREEN_KEY` / `settings_mqtt_screen` option from the settings menu when the feature is turned off, consistent with the other feature flags.

**Review by harshsomankar123-tech — CHANGES_REQUESTED** [review]
Thanks, @Paras-ydv. I only gave this a quick first pass, but I’ll do a full review ASAP. PTAL at the comments in the meantime.

File: catroid/src/main/AndroidManifest.xml (hunk 2 above)
Comment (Paras-ydv) [reply catroid/src/main/AndroidManifest.xml]: After reviewing the Paho documentation, all functionality required for the MQTT manager ticket can be implemented using `MqttAsyncClient` from `org.eclipse.paho.client.mqttv3`. We don't currently have a requirement for a background Android Service, so the additional Android wrapper does not provide a clear benefit for Catroid's use case.

I'll update the PR accordingly. Thanks!!

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java
Comment (Paras-ydv) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java:217]: Adding a FEATURE_MQTT_ENABLED check in SettingsFragment and removing the MQTT settings screen when the feature flag is disabled (consistent with the existing feature-gated settings entries) in the new commit.

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java [hunk 4]
```diff
@@ -408,6 +431,38 @@ public static boolean isRaspiSharedPreferenceEnabled(Context context) {
 		return getBooleanSharedPreference(false, SETTINGS_SHOW_RASPI_BRICKS, context);
 	}
 
+	public static boolean isMqttSharedPreferenceEnabled(Context context) {
+		return getBooleanSharedPreference(false, SETTINGS_SHOW_MQTT_BRICKS, context);
+	}
+
+	public static String getMqttHost(Context context) {
+		return getSharedPreferences(context).getString(MQTT_HOST, "");
```
Comment (harshsomankar123-tech) [inline catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java]: @Paras-ydv Rather than setting the default host to an empty string (which causes the unit tests to fail), we can move the default IP address to strings.xml and load it using context.getString(R.string.default_mqtt_host) in SettingsFragment.java.

This also resolves the SonarCloud hardcoded IP warning, since IP addresses in XML resources aren't flagged, while preserving the default behavior and keeping all unit tests passing.

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java [hunk 5]
```diff
@@ -131,6 +131,17 @@
...
 
+	public static final String MQTT_SCREEN_KEY = "settings_mqtt_screen";
+	public static final String SETTINGS_SHOW_MQTT_BRICKS = "setting_mqtt_bricks";
+	public static final String MQTT_CONNECTION_SETTINGS_CATEGORY = "setting_mqtt_category";
+	public static final String MQTT_HOST = "setting_mqtt_host";
+	public static final String MQTT_PORT = "setting_mqtt_port";
+	public static final String MQTT_TLS = "setting_mqtt_tls";
+	public static final String MQTT_USERNAME = "setting_mqtt_username";
+	public static final String MQTT_PASSWORD = "setting_mqtt_password";
+	public static final String MQTT_CLIENT_ID = "setting_mqtt_client_id";
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java:142]: solve it

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java (hunk 4 above)
Comment (Paras-ydv) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java]: thanks and nice idea!!

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java [hunk 6]
```diff
@@ -258,6 +259,15 @@ public void rasPiSettingsTest() {
 		checkPreference(R.string.preference_title_enable_raspi_bricks, SETTINGS_SHOW_RASPI_BRICKS);
 	}
 
+	@Category({Cat.AppUi.class, Level.Smoke.class, Cat.Gadgets.class})
+	@Test
+	public void mqttSettingsTest() {
+		onData(PreferenceMatchers.withTitle(R.string.preference_title_enable_mqtt_bricks))
+				.perform(click());
+
+		checkPreference(R.string.preference_title_enable_mqtt_bricks, SETTINGS_SHOW_MQTT_BRICKS);
```
Comment (harshsomankar123-tech) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java]: Use R.string.preference_title_mqtt_enabled instead of R.string.preference_title_enable_mqtt_bricks here. becoz the current string is the screen title (“MQTT extension”), not the actual checkbox preference inside MqttSettingsFragment, so Espresso won’t find a clickable checkbox.see the test is failing:

Image

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        bindSummaryToValue(findPreference(MQTT_CLIENT_ID) as EditTextPreference)
+
+        val portPreference = findPreference(MQTT_PORT) as EditTextPreference
+        portPreference.summary = portPreference.text
+        portPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
+            if (!isValidPort(newValue.toString())) {
+                Toast.makeText(activity, R.string.preference_mqtt_port_invalid, Toast.LENGTH_SHORT).show()
+                return@OnPreferenceChangeListener false
+            }
+            portPreference.summary = newValue.toString()
```
Comment (harshsomankar123-tech) [inline catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt:80]: Is there a reason we are adding a listener here that just returns true?

**Review by harshsomankar123-tech — CHANGES_REQUESTED** [review]
@Paras-ydv PTAL the review and questions!
Thanks!

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt (hunk 7 above)
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt:80]: I mean since EditTextPreference automatically persists values to SharedPreferences by default, a custom change listener that only returns true is redundant. plz erfer password preference listener on lines 80–81

**Review by wslany — CHANGES_REQUESTED** [review]
Many thanks!!

MQTT password is stored in default SharedPreferences
[mqtt_preferences.xml (line 63)](W:/Codex/Catroid-pr-5220-review/catroid/src/main/res/xml/mqtt_preferences.xml:63) defines the broker password as an EditTextPreference, and [SettingsFragment.java (line 456)](W:/Codex/Catroid-pr-5220-review/catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/SettingsFragment.java:456) reads it from default shared preferences. That means broker credentials are persisted in plaintext app prefs. Since Catroid already has androidx.security:security-crypto and an EncryptedSharedPreferences pattern, this should either move to encrypted storage or avoid persisting the password.

MQTT Espresso smoke test toggles the navigation entry instead of the actual checkbox
In [SettingsFragmentTest.java (line 264)](W:/Codex/Catroid-pr-5220-review/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java:264), the first click opens the MQTT settings screen using the title preference_title_enable_mqtt_bricks (“MQTT extension”). Then [line 268 (line 268)](W:/Codex/Catroid-pr-5220-review/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java:268) passes the same title into checkPreference, whose helper expects a checkbox row to toggle. Inside the MQTT screen, the checkbox title is preference_title_mqtt_enabled (“Enable MQTT bricks”), so the test either fails to find the row or does not verify SETTINGS_SHOW_MQTT_BRICKS. It should navigate once, then call checkPreference(R.string.preference_title_mqtt_enabled, SETTINGS_SHOW_MQTT_BRICKS).

**Comment by Paras-ydv** [comment]
Hi @harshsomankar123-tech and @wslany,

I've addressed the review comments:

- Updated the Espresso test to use R.string.preference_title_mqtt_enabled for the actual checkbox in MqttSettingsFragment.
- Removed the redundant password preference listener that only returned true.
- Changed MQTT password storage to use EncryptedSharedPreferences instead of the default SharedPreferences, and updated SettingsFragment.getMqttPassword() accordingly.

PTAL. Thanks!

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java (hunk 6 above)
Comment (Paras-ydv) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/SettingsFragmentTest.java]: done

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt (hunk 7 above)
Comment (Paras-ydv) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt:80]: changed in new commit. PTAL.

File: catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            true
+        }
+
+        val passwordPreference = findPreference(MQTT_PASSWORD) as EditTextPreference
+        passwordPreference.isPersistent = false
+        passwordPreference.text = encryptedPrefs.getString(MQTT_PASSWORD, "")
+        passwordPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
+            with(encryptedPrefs.edit()) {
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/ui/settingsfragments/MqttSettingsFragment.kt]: @Paras-ydv solve that lint issue!

**Comment by Paras-ydv** [comment]
Hi @wslany and @harshsomankar123-tech ,

I've implemented the MqttPasswordRepository approach we discussed.

Here's a summary of what was done:

- Added `MqttPasswordRepository` interface with `getPassword()`, `setPassword()`, and `clearPassword()`
- `DefaultMqttPasswordRepository` wraps `EncryptedSharedPreferences` (lazy-initialised), following the same pattern as `LocalHashVersionRepository`
- Registered as a Koin singleton in `repositoryModules` inside `CatroidKoinHelper`
- `MqttSettingsFragment` injects it via `by inject()` - password preference is non-persistent and reads/writes through the repository
- `SettingsFragment.getMqttPassword(Context)` retains its existing signature but delegates to the injected repository internally, so no other callers needed to change
- Unit tests mock the interface via a Koin test module (no `mockkStatic`, no Android Keystore dependency) and use `verify(exactly = 1)` to assert both the read and write paths

Let me know if you'd like any adjustments!!!

File: catroid/src/test/java/org/catrobat/catroid/test/mqtt/MqttSettingsTest.kt [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    fun testMqttPasswordPersistsAfterWrite() {
+        every { mockMqttPasswordRepository.getPassword() } returns "secret"
+        assertEquals("secret", SettingsFragment.getMqttPassword(context))
+        verify(exactly = 1) { mockMqttPasswordRepository.getPassword() }
+    }
+
+    @Test
+    fun testMqttPasswordSetPasswordDelegatesToRepository() {
+        mockMqttPasswordRepository.setPassword("secret")
+        verify(exactly = 1) { mockMqttPasswordRepository.setPassword("secret") }
```
Comment (harshsomankar123-tech) [inline catroid/src/test/java/org/catrobat/catroid/test/mqtt/MqttSettingsTest.kt]: This test case is redundant since it only calls setPassword directly on the mock and verifies the call on the mock itself, without exercising any production code. We should either update it to test the actual fragment's preference change listener, or remove this test case entirely.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/repository/MqttPasswordRepository.kt [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import org.catrobat.catroid.ui.settingsfragments.SettingsFragment.MQTT_ENCRYPTED_PREFS
+import org.catrobat.catroid.ui.settingsfragments.SettingsFragment.MQTT_PASSWORD
+
+interface MqttPasswordRepository {
+    fun getPassword(): String
+    fun setPassword(password: String)
+    fun clearPassword()
+}
+
+class DefaultMqttPasswordRepository(context: Context) : MqttPasswordRepository {
```
Comment (harshsomankar123-tech) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/repository/MqttPasswordRepository.kt]: Since DefaultMqttPasswordRepository relies on EncryptedSharedPreferences (which interacts with the Android Keystore), we should add a concrete unit test for it using a real Robolectric context to ensure it stores and retrieves values correctly without throwing any security or initialization exceptions. Currently, the repository is only mocked in the tests.

**Review by harshsomankar123-tech — CHANGES_REQUESTED** [review]
@Paras-ydv Thanks! Sorry for the delay. I was actually a bit busy working on a new feature for the organization. I just have a few final comments; otherwise, this PR is good to go from my side.

File: catroid/src/test/java/org/catrobat/catroid/test/mqtt/MqttSettingsTest.kt (hunk 9 above)
Comment (Paras-ydv) [reply catroid/src/test/java/org/catrobat/catroid/test/mqtt/MqttSettingsTest.kt]: Thanks! Removed the redundant test as suggested. The repository is now covered by dedicated unit tests, and MqttSettingsTest only verifies the interaction with the mocked repository.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/repository/MqttPasswordRepository.kt (hunk 10 above)
Comment (Paras-ydv) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/repository/MqttPasswordRepository.kt]: Implemented this in the latest commit. I refactored `DefaultMqttPasswordRepository` to accept a `SharedPreferences` instance via dependency injection, with `EncryptedSharedPreferences` now created in `Koin`. This allows the repository to be tested using a regular `Robolectric SharedPreferences`, and I've added unit tests covering the default, set/get, and clear password scenarios.

**Review by harshsomankar123-tech — APPROVED** [review]
@Paras-ydv Thanks! LGTM
cc @wslany @reichli

**Review by wslany — APPROVED** [review]
